src/cls/rbd/cls_rbd.cc

```diff
@@ -3684,6 +3689,84 @@ int instances_remove(cls_method_context_t hctx, const string &instance_id) {
   return 0;
 }
 
+int mirror_image_map_list(cls_method_context_t hctx,
+                          const std::string &start_after,
+                          uint64_t max_return,
+                          std::map<std::string, cls::rbd::MirrorImageMap> *image_map) {
+  int max_read = MIN(RBD_MAX_KEYS_READ, max_return);
+  std::string last_read = mirror_image_map_key(start_after);
+
+  while (max_read) {
+    std::map<std::string, bufferlist> vals;
+    CLS_LOG(20, "last read: '%s'", last_read.c_str());
+
+    int r = cls_cxx_map_get_vals(hctx, last_read, MIRROR_IMAGE_MAP_KEY_PREFIX,
+                                 max_read, &vals);
+    if (r < 0) {
+      CLS_ERR("error reading image map: %s", cpp_strerror(r).c_str());
+      return r;
+    }
+
+    if (vals.empty()) {
+      return 0;
+    }
+
+    for (auto it = vals.begin(); it != vals.end(); ++it) {
+      const std::string& global_image_id =
+        it->first.substr(MIRROR_IMAGE_MAP_KEY_PREFIX.size());
+
+      cls::rbd::MirrorImageMap imagemap;
+      bufferlist::iterator iter = it->second.begin();
+      try {
+        ::decode(imagemap, iter);
+      } catch (const buffer::error& err) {
+        CLS_ERR("could not decode image map payload: %s", cpp_strerror(r).c_str());
+        return -EINVAL;
+      }
+
+      image_map->insert(std::make_pair(global_image_id, imagemap));
+    }
+
+    max_read = MIN(RBD_MAX_KEYS_READ, max_return - image_map->size());
+    if (!vals.empty()) {
+      last_read = mirror_image_map_key(image_map->rbegin()->first);
```

Comment:
The key should already be good-to-go -- no need to prefix an additional prefix onto it.